### PR TITLE
Add editorconfig

### PR DESCRIPTION
**Pull request by cuttlefish**
[EditorConfig](https://EditorConfig.org) is a cross editor/IDE configuration system. It's simple and is supported out-of-the-box by many editors and as plugins in most of the rest.

I generally don't like committing editor/IDE config files in repos, but I find this one useful.

I arbitrarily chose some settings that seemed to match what's currently in the repo. I'm happy to make modifications.

### CODE REVIEW

**Comment by gadomski** [comment]
This might be a Larger Discussion™, but aren't a lot of these things caught/checked by the Python linters, and don't most modern editors integrate linters? This feels anti-DRY to me, but again I might just be yelling at clouds.

**Comment by cuttlefish** [comment]
Yeah, it probably is a bigger conversation. My concern was less the Python files, but everything else. I'm seeing a bunch of PRs with messy line endings and such.
The only Python based Markdown linter I could find isn't maintained. To be fair, I didn't look for very long. Using non-Python linters will complicate setting up dev environments a fair bit. Not a big deal on the Docker side, but annoying for everyone else. I don't want to force people to run Docker for linting, but I guess it's a possibility.

It's really a cultural problem, not a technical one. But some people are just going to `git add . && git commit` and flood no-op white space commits all over the place, no matter what anyone else says. I was hoping editorconfig might at least mitigate the effects somewhat. Kind of a "defence in depth" strategy. But I'm definitely not wed to the idea and it doesn't actually solve the problem.

**Comment by gadomski** [comment]
> I was hoping editorconfig might at least mitigate the effects somewhat. Kind of a "defence in depth" strategy.

I'm into this for sure. Does it work w/o any setup (aka do editors just find it)?

File: . editorconfig [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+[make.bat]
+indent_style = tab
+
+[*.md]
+
+[*.py,setup.cfg]
+
+[coverage.xml]
+indent_style = tab
+
```
Comment (gadomski) [inline . editorconfig:26]: These shouldn't be checked in to the repo, correct? So we shouldn't need config for them.

File: . editorconfig [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+trim_trailing_whitespace = true
+indent_style = space
+indent_size = 4
+
+[Makefile]
+indent_style = tab
+
+[make.bat]
+indent_style = tab
+
```
Comment (gadomski) [inline . editorconfig:19]: Since we don't have any of these (and please never will, especially `make.bat`), should we include them at all?

**Review by gadomski — APPROVED** [review]
I'm good with adding this per your "defense in depth" strategy, makes sense to me. I put a couple comments in for files that we don't (and hopefully shouldn't) have checked in, which we could remove or keep -- I'm a minimalist so I'd nuke them, but they're really not hurting anyone :-).

**Comment by cuttlefish** [comment]
> Does it work w/o any setup (aka do editors just find it)?

A lot of editors support it natively (usually it acts as new defaults on whatever project you're editing), but unfortunately several of the most popular require plugins (vim, emacs, VS Code, Atom, Sublime Text).

Hmm... we could add something like [https://pypi.org/project/editorconfig-checker/](https://pypi.org/project/editorconfig-checker/) to the linting scripts. That would at least enforce it and would just be a `requirements-dev.txt` dependency.

File: . editorconfig (hunk 1 above)
Comment (cuttlefish) [reply . editorconfig:26]: Oops! You're 💯 right

File: . editorconfig (hunk 2 above)
Comment (cuttlefish) [reply . editorconfig:19]: Those (including `make.bat`) are generated by sphinx

Comment (gadomski) [reply . editorconfig:19]: Ah forgot about that, disregard.

**Review by gadomski — APPROVED** [review]
Good by me. Is the `.ecrc` required? Feels weird to have two dotfiles for one product. Also `editorconfig-checker` is >100MB :-(. Not a big deal, just noting that we're adding some weight here.

Screen Shot 2021-07-26 at 4 40 48 AM

**Comment by cuttlefish** [comment]
Yikes... 100MB was a lot more than I was expecting. Well, I'll merge this in for now.
I removed the `.ecrc` config file. It turns out that `editorconfig-checker` does the sane thing and by default only checks things that are added to the git repo.

**Pull request merged**